validator-network/src/network_impl.rs

```diff
@@ -76,50 +88,111 @@ where
+                    validator_id,
+                    error = ?e,
+                    %public_key,
+                    "Unable to resolve validator peer ID: Network error"
+                );
+                CacheState::Error(e)
+            }
+        };
+        self.validator_peer_id_cache
+            .write()
+            .insert(public_key.compressed().clone(), cache_value);
+    }
+
     /// Look up the peer ID for a validator ID.
     async fn get_validator_peer_id(
         &self,
         validator_id: u16,
-    ) -> Result<N::PeerId, NetworkError<N::Error>> {
-        let (peer_id, public_key) = {
-            let public_key = self
-                .validator_keys
-                .read()
-                .get(usize::from(validator_id))
-                .ok_or(NetworkError::UnknownValidator(validator_id))?
-                .clone();
+    ) -> Result<N::PeerId, NetworkError<N::Error>>
+    where
+        <N as Network>::Error: Sync + Send,
+    {
+        let public_key = self
+            .validator_keys
+            .read()
+            .get(usize::from(validator_id))
+            .ok_or(NetworkError::UnknownValidator(validator_id))?
+            .clone();
 
-            let peer_id = self
+        let update_peer_cache = {
+            if let Some(cache_state) = self
                 .validator_peer_id_cache
                 .read()
                 .get(&public_key.compressed().clone())
-                .cloned();
-            (peer_id, public_key)
+            {
+                match cache_state {
+                    CacheState::Resolved(peer_id) => return Ok(*peer_id),
+                    CacheState::Error(_) => true,
+                    CacheState::InProgress => {
+                        log::debug!(validator_id, "Record resolution is in progress");
+                        false
+                    }
+                }
+            } else {
+                log::error!(
+                    ?public_key,
+                    validator_id,
+                    "Could not find peer ID for validator in DHT",
+                );
+                true
+            }
         };
-
-        if let Some(peer_id) = peer_id {
-            Ok(peer_id)
-        } else if let Some(peer_id) = Self::resolve_peer_id(&self.network, &public_key).await? {
-            self.validator_peer_id_cache
-                .write()
-                .insert(public_key.compressed().clone(), peer_id);
-            Ok(peer_id)
-        } else {
-            log::error!(
-                "Could not find peer ID for validator in DHT: public_key = {:?}",
-                public_key
-            );
-            Err(NetworkError::UnknownValidator(validator_id))
+        if update_peer_cache {
+            // Cache is empty for this validator ID, query the entry
+            {
+                self.validator_peer_id_cache
+                    .write()
+                    .insert(public_key.compressed().clone(), CacheState::InProgress);
```

Comment:
In order not to run into concurrency issues, I think the `match cache_state` from above needs to be repeated here, with the write lock present. Otherwise, another future might have inserted a `CacheState::InProgress` or even a `CacheState::Resolved` here, and we'd start a second future or throw away the result.

Reply:
Yes, this should take an upgradable read in the beginning and keep it. Upgrade here then.